findcoordinator.go

```diff
@@ -2,8 +2,120 @@ package kafka
 
 import (
 	"bufio"
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/segmentio/kafka-go/protocol/findcoordinator"
+)
+
+// CoordinatorKeyType is used to specify the type of coordinator to look for
+type CoordinatorKeyType int8
+
+const (
+	// CoordinatorKeyTypeConsumer type is used when looking for a Group coordinator
+	CoordinatorKeyTypeConsumer CoordinatorKeyType = 0
+
+	// CoordinatorKeyTypeTransaction type is used when looking for a Transaction coordinator
+	CoordinatorKeyTypeTransaction CoordinatorKeyType = 1
 )
 
+// FindCoordinatorRequest is the request structure for the FindCoordinator function
+type FindCoordinatorRequest struct {
+	// Address of the kafka broker to send the request to.
+	Addr net.Addr
+
+	// The coordinator key.
+	Key string
+
+	// The coordinator key type. (Group, transaction, etc.)
+	KeyType CoordinatorKeyType
+}
+
+// FindCoordinatorResponseCoordinator contains details about the found coordinator
+type FindCoordinatorResponseCoordinator struct {
+	// NodeID holds the broker id.
+	NodeID int32
+
+	// Host of the broker
+	Host string
+
+	// Port on which broker accepts requests
+	Port int32
+}
+
+// FindCoordinatorResponse is the response structure for the FindCoordinator function
+type FindCoordinatorResponse struct {
+	// The Transaction/Group Coordinator details
+	Coordinator *FindCoordinatorResponseCoordinator
+
+	// The amount of time that the broker throttled the request.
+	Throttle time.Duration
+
+	// An error that may have occurred while attempting to retrieve Coordinator
+	//
+	// The error contains both the kafka error code, and an error message
+	// returned by the kafka broker.
+	Error error
+
+	// Error code
+	ErrorCode int
```

Comment:
Since we already inject the error in the `Error` field, this field seems redundant?
```suggestion
```

Reply:
The current `makeError` implementation returns an `error` object from the errorCode provided.
Once that is done, the error code information can not be accessed again. There should be a way to access error code information as well to ensure downstream functions that make use of `FindCoordinator` have more visibility. 

I will refactor this to still retain error code information.

Reply:
I don't think that's accurate, the error returned by `makeError` is either a `kafka.Error`, or wraps one.

The underlying `kafka.Error` can always be retrieved with:
```go
var kafkaError kafka.Error
if errors.As(err, &kafkaErr) {
  // use kafkaError
  ...
}
```

Reply:
Interesting. TIL 😮  
Will make the necessary modification.